Update to EoS fitting

bobmyhill opened the pull request:
This update is intended to improve the EoS fitting and plotting routines in burnman. Improvements include:
- Full propagation of P, T, property covariance matrices via non-linear least squares 
- Calculation of uncertainties
- Corner plots for covariances
- Delta method estimation of confidence and prediction bands for volume/pressure/temperature
- Confidence and prediction bands for other physical properties
- Examples for P-V, P-T-V and P-T-V-H EoS fitting
- Implementation of extreme value theory for data removal

Still to do:
- [x] Document functions
- [x] Create corner plot to illustrate covariances.
- [x] Check (and possibly restructure) the delta method for other properties (to look at, for example, the constraints on P-wave velocity or gibbs free energy.
- [x] Fix tests
- [x] Generalise PTV function to PT-property, where property is an attribute of a mineral (e.g. 'shear_wave_velocity')
- [x] Generalise nonlinear_least_squares_fit() to find the best fit model using more than one type of function, with flags for each datum to tell model.function(x, flag) and model.normal(x, flag) which function to use.

Comments and suggestions on code structure/further additions welcome.

tjhei commented on burnman/nonlinear_fitting.py:
```diff
@@ -0,0 +1,314 @@
+            model.jacobian[:,prm_i] = (residual_arr_1 - residual_arr_0)/(2.*diag_delta[prm_i][prm_i])
+        model.set_params(param_values) # reset params
+        
+        return None
```
remove?

tjhei commented on burnman/nonlinear_fitting.py:
```diff
@@ -0,0 +1,314 @@
+    
+    n_it = -1
+    tol_achieved = False
+    while tol_achieved == False and n_it < max_lm_iterations:
```
I would have written a ``for`` loop with a ``break``

tjhei commented on burnman/nonlinear_fitting.py:
```diff
@@ -0,0 +1,314 @@
+
+
+def attribute_function(mineral, attributes, powers=[]):
+    if type(attributes) is str:
```
can you please document this?

tjhei commented on burnman/nonlinear_fitting.py:
```diff
@@ -0,0 +1,314 @@
+                                verbose = False):
+
+    """
+    Function to compute the "best-fit" parameters for a model
```
you are replacing ``scipy.optimize.curve_fit`` by this. Can you explain what you are doing differently now? Is this still a Levenberg-Marquardt?

bobmyhill replied:
Yes, this is still L-M. The new function runs a full nonlinear inversion with covariances in all parameters, rather than an inversion with errors only in one direction ("y"). No particular coordinate axis is special in the new formulation, so from that perspective it's a bit like orthogonal distance regression.

tjhei commented on burnman/nonlinear_fitting.py:
```diff
@@ -0,0 +1,314 @@
+        Print some information to standard output
+        
+
+    Attributes added to model
```
very cool design!

bobmyhill replied:
Thanks!

bobmyhill replied:
Coming from you, that means a lot :)

bobmyhill commented:
Any thoughts on what to do with examples/example_fit_data.py? The fitting can be done better now (see example_tools.py and example_eos_fitting.py), but the incorrect and correct averaging example is still a useful thing to have.

bobmyhill commented:
@tjhei I think example_eos_fitting.py should serve as a decent starting point for the web tool. I need to think a bit about how the plots should look for the T-V, P-V and P-V-T cases, but the core functions are all there now.

bobmyhill commented:
@sannecottaar Any thoughts on what to do with examples/example_fit_data.py? he incorrect and correct averaging example is still a useful thing to have.

sannecottaar commented:
I think any examples can be modified and updated. The incorrect
extrapolation example is covered in the scripts that are linked to the
paper, so no need to keep it here too.

On Sun, Feb 5, 2017 at 1:02 AM, Bob Myhill <notifications@github.com> wrote:

> @sannecottaar <https://github.com/sannecottaar> Any thoughts on what to
> do with examples/example_fit_data.py? he incorrect and correct averaging
> example is still a useful thing to have.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/geodynamics/burnman/pull/273#issuecomment-277489701>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGJs2kPwcxgd3Fnjrsr1F14IsO25L6a4ks5rZR_DgaJpZM4L28fu>
> .
>

bobmyhill commented:
@sannecottaar Ok, great, thanks. I've updated (and renamed) the example, and generalised the fitting function to be P-T-parameter (where the user chooses the parameter).

tjhei commented:
> Any thoughts on what to do with examples/example_fit_data.py?

I agree with Sanne. The one from the paper is still in ``contrib/CHRU2014/paper_fit_data.py``

sannecottaar commented on examples/example_fit_eos.py:
```diff
@@ -0,0 +1,328 @@
+    norm = colors.BoundaryNorm(bounds, cmap.N)
+    
+    im = ax.scatter(PTV_data[:,0]/1.e9, PTV_data[:,1], c=fitted_eos.weighted_residuals[V_mask], cmap=cmap, norm=norm, s=50)
+    fig.colorbar(im, ax=ax)
```
Super cool stuff!!
Minor comment here, this plot needs some labels (and the same one below) .

bobmyhill replied:
Good catch Sanne! I've tweaked the example to be closer to the neat version in contrib/eos_fitting. All the figures should now have sensible axes.

bobmyhill commented:
@tjhei What's going on with the occasional tester failures for example_build_planet.py?

bobmyhill commented:
Everyone happy with this lil guy?

tjhei commented on burnman/data/input_fitting/Dubrovinsky_Saxena_1997_TV_periclase.dat:
```diff
@@ -0,0 +1,34 @@
+3015	12.9723	0.025
+# Dubrovinsky, L. S., and S. K. Saxena.
+# Thermal expansion of periclase (MgO) and tungsten (W) to melting temperatures.
+# Physics and Chemistry of Minerals 24.8 (1997): 547-550.
```
add missing newline

bobmyhill commented:
Ok, newlines added at the end of the four new data files

bobmyhill commented on contrib/eos_fitting/test.dat:
```diff
@@ -0,0 +1,94 @@
+H	1.000000e+05	3.731500e+02	3.863924e+03	0.000000e+00	1.000000e+00	1.545570e+00
```
Good catch, thanks @tjhei 
Comments added.
